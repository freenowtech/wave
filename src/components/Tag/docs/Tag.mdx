---
name: Tag
menu: Components
route: /components/tag
---

import { Playground } from 'docz';
import { Tag } from '../Tag.tsx';
import { DismissibleTag } from './DismissibleTag.tsx';
import { DismissableTagPropsTable } from './DismissableTagPropsTable.tsx';

# Tag

The Tag component is used to categorize items. The text inside the Tag should be kept as short as possible. Multi-labels
should be avoided as well.

## Properties

<DismissableTagPropsTable />
<br/>

<DismissibleTag />

```
<Tag>Filter</Tag>
```

## Playground

<Playground>
    <Tag>Hive</Tag>
    <Tag dismissible={false}>Taxi</Tag>
</Playground>
