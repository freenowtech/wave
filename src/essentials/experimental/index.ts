export * from './theme';
export { createGlobalStyle, ColorPalette } from './Colors';
export { darkTheme } from './globalStyles';
export { getSemanticValue, getBareValue, getBareHslValue, getSemanticHslValue } from './cssVariables';
