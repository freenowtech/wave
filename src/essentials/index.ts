export { Elevation } from './Elevation/Elevation';
export { Spaces } from './Spaces/Spaces';
export { Breakpoints, MediaQueries } from './Breakpoints/Breakpoints';
export type { ReadCssColorVariable, ReadCssVariable } from './Colors/types';
export { GlobalStyle as ClassicColors } from './Colors/Colors';
export { GlobalStyle as ModernColors } from './Colors/ModernColors';
