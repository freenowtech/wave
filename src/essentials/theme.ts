import { Colors, SemanticColors } from './Colors/Colors';
import { Spaces } from './Spaces/Spaces';
import { Breakpoints, MediaQueries } from './Breakpoints/Breakpoints';

const theme = {
    breakpoints: Breakpoints,
    colors: Colors,
    // todo: rename semanticColors to colors in the next major release
    semanticColors: SemanticColors,
    fontSizes: ['0.75rem', '0.875rem', '1rem', '1.25rem', '1.5rem', '2rem', '2.5rem', '3rem'],
    fontWeights: {
        normal: 400,
        semibold: 600,
        bold: 700
    },
    fonts: {
        normal: fontStack(['Open Sans', 'sans-serif']),
        monospace: fontStack(['Hack', 'monospace'])
    },
    mediaQueries: MediaQueries,
    space: Spaces,
    radii: ['0rem', '0.125rem', '0.25rem', '0.5rem'],
    iconSizes: {
        small: 12,
        medium: 24,
        large: 28
    },
    shadows: {
        small: '0 0.0625rem 0.25rem 0 rgba(0,0,0,0.1)',
        medium: '0 0 0.5rem 0.1875rem rgba(0,0,0,0.08)',
        large: '0 0 0.75rem 0.25rem rgba(0,0,0,0.12)'
    }
};

/**
 * Custom aliases declaration
 * https://styled-system.com/theme-specification/#scale-aliases
 */

// @ts-ignore
theme.fontSizes.small = theme.fontSizes[1];
// @ts-ignore
theme.fontSizes.medium = theme.fontSizes[2];
// TODO: the alias large should match with scale on theme "fontSizes" but it may require a breaking change
// @ts-ignore
theme.fontSizes.large = '1.125rem';

function fontStack(fonts) {
    return fonts.map(font => (font.includes(' ') ? `"${font}"` : font)).join(', ');
}

export { theme };
