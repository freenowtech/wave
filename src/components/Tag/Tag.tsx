import React, { FC, MouseEvent, PropsWithChildren } from 'react';
import styled from 'styled-components';
import { margin, MarginProps } from 'styled-system';

import { Colors } from '../../essentials';
import { theme } from '../../essentials/theme';
import { CloseIcon } from '../../icons';
import { get } from '../../utils/themeGet';
import { Text } from '../Text/Text';

interface TagProps extends MarginProps {
    /**
     * The function to handle when the dismiss icon is clicked
     */
    onDismiss?: (e: MouseEvent) => void;
    /**
     * The prop to determine whether the dismiss functionality is enabled
     */
    dismissible?: boolean;
}

const TagText = styled(Text).attrs({ theme })<Pick<TagProps, 'dismissible'>>`
    color: ${Colors.ACTION_BLUE_900};
    margin-left: 0.75rem;
    margin-right: ${props => (props.dismissible ? '0.25rem' : '0.75rem')};
    font-size: ${get('fontSizes.1')};
    font-weight: ${get('fontWeights.semibold')};
    cursor: default;
    transition: color 125ms ease;
`;

const DismissIcon = styled(CloseIcon).attrs({ size: 18 })`
    margin-right: 0.5rem;
    transition: color 125ms ease;

    &:hover {
        cursor: pointer;
    }
`;

const TagWrapper = styled.div.attrs({ theme })<TagProps>`
    box-sizing: border-box;
    background-color: ${Colors.ACTION_BLUE_50};
    border: solid 0.0625rem ${Colors.ACTION_BLUE_900};
    display: inline-flex;
    align-items: center;
    border-radius: 2rem;
    height: 2rem;
    margin-right: 0.25rem;
    margin-bottom: 0.375rem;
    padding: 0.375rem 0;
    transition: background-color 125ms ease;

    ${margin}

    &:hover {
        background-color: ${Colors.ACTION_BLUE_900};

        > ${TagText} {
            color: ${Colors.WHITE};
        }

        > ${DismissIcon} {
            color: ${Colors.WHITE};
        }
    }
`;

const Tag: FC<TagProps> = ({ children, onDismiss, dismissible = true, ...rest }: PropsWithChildren<TagProps>) => (
    <TagWrapper {...rest}>
        <TagText dismissible={dismissible}>{children}</TagText>
        {dismissible && <DismissIcon data-testid="dismiss-icon" color={Colors.ACTION_BLUE_900} onClick={onDismiss} />}
    </TagWrapper>
);

export { Tag, TagProps };
