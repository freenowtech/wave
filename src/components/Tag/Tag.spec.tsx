import { render, fireEvent, screen } from '@testing-library/react';
import * as React from 'react';
import { Tag } from './Tag';
import { SemanticColors } from '../../essentials';

describe('Tag', () => {
    it('renders and can be dismissed', () => {
        const onDismiss = jest.fn();
        const { getByTestId } = render(<Tag onDismiss={onDismiss}>Lorem</Tag>);

        fireEvent.click(getByTestId('dismiss-icon'));

        expect(onDismiss).toBeCalledTimes(1);
    });

    it('renders and can not be dismissed when dismissible is set to false', () => {
        const onDismiss = jest.fn();
        const { queryByTestId } = render(
            <Tag onDismiss={onDismiss} dismissible={false}>
                Lorem
            </Tag>
        );

        expect(queryByTestId('dismiss-icon')).toBeNull();
    });

    it('renders disabled variant', () => {
        const { container } = render(<Tag variant="disabled">Lorem</Tag>);

        expect(container.firstChild).toHaveStyle(`
            border-color: ${SemanticColors.border.primary};
        `);
        expect(screen.getByText('Lorem')).toHaveStyle(`
            color: ${SemanticColors.text.disabled};
        `);
        expect(screen.getByTestId('dismiss-icon')).toHaveStyle(`
            color: ${SemanticColors.icon.disabled};
        `);
    });

    it('renders error variant', () => {
        const { container } = render(<Tag variant="error">Lorem</Tag>);

        expect(container.firstChild).toHaveStyle(`
            background-color: ${SemanticColors.background.danger};
            border-color: ${SemanticColors.border.dangerEmphasized};
        `);
        expect(screen.getByText('Lorem')).toHaveStyle(`
            color: ${SemanticColors.text.dangerInverted};
        `);
        expect(screen.getByTestId('dismiss-icon')).toHaveStyle(`
            color: ${SemanticColors.icon.danger};
        `);
    });
});
