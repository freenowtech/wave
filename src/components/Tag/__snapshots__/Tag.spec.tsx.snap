// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag renders with default props 1`] = `
.c2 {
  color: #001E3E;
  font-size: 1rem;
  font-family: "Open Sans",sans-serif;
  line-height: 1.4;
  margin: 0;
}

.c3 {
  margin-left: 0.75rem;
  margin-right: 0.25rem;
  font-size: 0.875rem;
  font-weight: 600;
  cursor: default;
  -webkit-transition: color 125ms ease;
  transition: color 125ms ease;
}

.c5 {
  margin-right: 0.5rem;
  -webkit-transition: color 125ms ease;
  transition: color 125ms ease;
}

.c5:hover {
  cursor: pointer;
}

.c0 {
  box-sizing: border-box;
  border: solid 0.0625rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 2rem;
  height: 2rem;
  margin-right: 0.25rem;
  margin-bottom: 0.375rem;
  padding: 0.375rem 0;
  -webkit-transition: background-color 125ms ease;
  transition: background-color 125ms ease;
  background-color: #F1F7FD;
  border-color: #096BDB;
}

.c0 > .c1 {
  color: #096BDB;
}

.c0 > .c4 {
  color: #096BDB;
}

.c0:hover {
  background-color: #096BDB;
  border-color: #096BDB;
}

.c0:hover > .c1 {
  color: #FFFFFF;
}

.c0:hover > .c4 {
  color: #FFFFFF;
}

<div
  class="c0"
>
  <span
    class="c1 c2 c3"
  >
    Lorem
  </span>
  <svg
    class="c4 c5"
    color="#096BDB"
    data-testid="dismiss-icon"
    height="18"
    viewBox="0 0 24 24"
    width="18"
  >
    <path
      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      fill="currentColor"
      fill-rule="nonzero"
    />
  </svg>
</div>
`;
